### DISCUSSION TITLE
Revolving errors in qmd with robservable and {ojs} cells

### DISCUSSION THREAD

**Discussion by allisonhorst** [quarto-dev/quarto-cli#5344] [Q&A]
Hi everybody! [UPDATE: solved - using both robservable and {ojs} cells in a .qmd created conflicts!]

In a Quarto doc I'm trying to show different ways that folks can pull in content from Observable notebooks to a Quarto doc, or work in {ojs} cells. Repo: [https://github.com/allisonhorst/robservable-testing]()

_Sometimes_ it shows up perfectly, but more often I get errors throughout - and the error messages will sometimes change when I refresh the page. For example, below are four different outcomes for the same cell, only refreshing the page between: 

<img width="448" alt="Screenshot 2023-04-28 at 8 13 44 AM" src="https://user-images.githubusercontent.com/25372780/235186282-96ea5827-f68b-430b-a298-90f028634869.png">
<img width="441" alt="Screenshot 2023-04-28 at 8 13 20 AM" src="https://user-images.githubusercontent.com/25372780/235186300-12decb99-1175-4d48-8bcf-2b1767861a01.png">
<img width="444" alt="Screenshot 2023-04-28 at 8 12 56 AM" src="https://user-images.githubusercontent.com/25372780/235186316-951f58c4-9d50-41ce-afa2-529a6165f688.png">
<img width="444" alt="Screenshot 2023-04-28 at 8 14 45 AM" src="https://user-images.githubusercontent.com/25372780/235186414-bb8e4322-61cf-4c03-ab37-c5a6130eacbe.png">

My coworker pointed me to this discussion (which is a bit over my head, to be honest): https://talk.observablehq.com/t/quarto-module-require/6573 (UPDATE): They also shared an idea: "robservable conflicts with the quarto-ojs-runtime because they both package d3-require" -- so this could be an issue with robservable?

Any ideas / recommendations for making this thing work consistently? Am I missing something important that I need in my Quarto doc? Thank you so much! 

I'm using: 

- RStudio Version 2023.05.0-daily+313 (2023.05.0-daily+313)
- Quarto Version 1.3.340

**Comment by allisonhorst** [comment]
OK some follow up: once I broke the notebook up with the robservable example separate from the {ojs} examples, then this works great! So it does seem to be a conflict between the two. I'm gonna close this and follow up with the robservable developers.

**Reply by cscheid** [reply]
I don't know how robservable works, but I would be very surprised if it interacted well (or at all) with our implementation!

> They also shared an idea: "robservable conflicts with the quarto-ojs-runtime because they both package d3-require" -- so this could be an issue with robservable?

This is true, and in fact worse than you might expect: in order for our implementation to work well with jupyter widgets, we run a fork of `d3-require`. So it's almost certainly the case that if you run both robservable and quarto-ojs-runtime on the same page, they're going to step on each other's toes :(

Are there robservable features that don't exist in our `ojs` engine support that you'd need to use?

I should note that if you use robservable by itself on a page without an `ojs` code cell, then things _should_ work, because we only include quarto-ojs-runtime on pages we're positive require quarto-ojs-runtime.

**Reply by allisonhorst** [reply]
Thanks for following up, I really appreciate it! This is really helpful. 

I don't think I have any real-world need for robservable + ojs in a single .qmd except for this one case where I was trying to show the different ways that people can work between Observable and Quarto -- but splitting it up works fine! 

Thanks again for all of your work on Quarto and the ojs engine, it's really awesome and I appreciate your thoughtful response.

**Reply by allisonhorst** [reply]
Also just in case it's of interest, my awesome coworker Fabian shared this notebook and said it might be a useful alternate workaround because it "prevents another instance from grabbing window.define": https://observablehq.com/@mootari/require-conflict

**Reply by cscheid** [reply]
I'm going to have to study that notebook more carefully! Our issue is not (only, maybe?) with multiple instances of d3-require, but with how d3-require doesn't play nicely with Jupyter widget's expectation of the js environment. Our specific fork is [here](https://github.com/quarto-dev/quarto/tree/main/packages/ojs/external-d3-d3-require) but because it's deep inside a bunch of git subtree nonsense, it's hard to diff against the upstream, I apologize.
